### PR TITLE
Avoid the `if(ReceiveAsync) {await WaitAsync();}` pattern

### PR DESCRIPTION
**Pull request by mgravell**
Note this also refactors `Signal` from being behaving  like `AutoResetEvent`, to instead behaving  like `ManualResetEvent`

### CODE REVIEW

File: src/Channels.Networking.Sockets/Internal/SocketExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    {
+        /// <summary>
+        /// Note that this presumes that args.UserToken is already a Signal, and that args.Completed
+        /// knows to call .Set on the Signal
+        /// </summary>
+        public static Signal ReceiveSignalAsync(this Socket socket, SocketAsyncEventArgs args)
+        {
+            var signal = (Signal)args.UserToken;
+            signal.Reset();
+            if(!socket.ReceiveAsync(args))
```
Comment (davidfowl) [inline src/Channels.Networking.Sockets/Internal/SocketExtensions.cs]: nit: ctrl + k, ctrl + d

Comment (mgravell) [reply src/Channels.Networking.Sockets/Internal/SocketExtensions.cs]: c4ec35741031e45f5f9d716ecc463c8eb72f317e

File: src/Channels.Networking.Sockets/Internal/Signal.cs [hunk 2]
```diff
@@ -9,69 +9,62 @@ namespace Channels.Networking.Sockets.Internal
...
-        private volatile bool _isCompleted;
+        private static readonly Action _completedSentinel = delegate { };
 
         public Signal(ContinuationMode continuationMode = ContinuationMode.Synchronous)
         {
             _continuationMode = continuationMode;
         }
 
-        public bool IsCompleted => _isCompleted;
+        public bool IsCompleted => ReferenceEquals(Volatile.Read(ref _continuation), _completedSentinel);
```
Comment (benaadams) [inline src/Channels.Networking.Sockets/Internal/Signal.cs]: Probably not important, but I'd swap these so the Volatile.Read is last thing done, would reduce the window for false negatives by a few nano seconds 😉

Comment (mgravell) [reply src/Channels.Networking.Sockets/Internal/Signal.cs]: Sure, I can swap.

Comment (mgravell) [reply src/Channels.Networking.Sockets/Internal/Signal.cs]: cb26d5b2333afaaa8a2e1bb2fc4e86937f33e869

**Pull request merged**